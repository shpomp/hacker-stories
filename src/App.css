.App {
	height: 100vh;
	background-color: rgb(157, 157, 219);
	padding: 90px;
}
input {
	background-color: rgb(210, 210, 255);
	padding: 10px 10px;
	line-height: 20px;
	border: none;
}

.App-logo {
	height: 40vmin;
	pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
	.App-logo {
		animation: App-logo-spin infinite 20s linear;
	}
}

.App-header {
	background-color: #282c34;
	min-height: 100vh;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	font-size: calc(10px + 2vmin);
	color: white;
}

.App-link {
	color: #61dafb;
}
