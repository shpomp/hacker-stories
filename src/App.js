import styles from "./App.module.css";
import styled from "styled-components";
import axios from "axios";
import { useState, useEffect, useReducer, useCallback, useRef } from "react";
import List from "./components/List";
import SearchForm from "./components/SearchForm";
import Name from "./components/Name";

// ------- STYLED COMPONENTS -------

const StyledContainer = styled.div`
	height: 100vw;
	padding: 20px;
	background: #83a4d4;
	background: linear-gradient(to left, #b6fbff, #83a4d4);
	color: #171212;
`;
const StyledHeadlinePrimary = styled.h1`
	font-size: 48px;
	font-weight: 300;
	letter-spacing: 2px;
`;

// ------- APP -------

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

const useSemiPersistentState = (key, initialState) => {
	const isMounted = useRef(false);
	const [value, setValue] = useState(localStorage.getItem(key) || initialState);
	useEffect(() => {
		if (!isMounted.current) {
			isMounted.current = true;
		} else {
			console.log("A");
			localStorage.setItem(key, value);
		}
	}, [value, key]);
	return [value, setValue];
};

const ACTIONS = {
	SET_STORIES: "SET_STORIES",
	REMOVE_STORY: "REMOVE_STORY",
	STORIES_FETCH_INIT: "STORIES_FETCH_INIT",
	STORIES_FETCH_SUCCESS: "STORIES_FETCH_SUCCESS",
	STORIES_FETCH_FAILURE: "STORIES_FETCH_FAILURE",
};

const storiesReducer = (state, action) => {
	switch (action.type) {
		case ACTIONS.STORIES_FETCH_INIT:
			return {
				...state,
				isLoading: true,
				isError: false,
			};
		case ACTIONS.STORIES_FETCH_SUCCESS:
			return {
				...state,
				isLoading: false,
				isError: false,
				data: action.payload,
			};
		case ACTIONS.STORIES_FETCH_FAILURE:
			return {
				...state,
				isLoading: false,
				isError: true,
			};
		case ACTIONS.REMOVE_STORY:
			return {
				...state,
				data: state.data.filter(
					(story) => action.payload.objectID !== story.objectID
				),
			};
		default:
			throw new Error();
	}
};

const App = () => {
	const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
	const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);

	const [stories, dispatchStories] = useReducer(storiesReducer, {
		data: [],
		isLoading: false,
		isError: false,
	});

	const handleFetchStories = useCallback(async () => {
		dispatchStories({ type: ACTIONS.STORIES_FETCH_INIT });

		try {
			const result = await axios.get(url);

			dispatchStories({
				type: ACTIONS.STORIES_FETCH_SUCCESS,
				payload: result.data.hits,
			});
		} catch {
			dispatchStories({ type: ACTIONS.STORIES_FETCH_FAILURE });
		}
	}, [url]);

	useEffect(() => {
		handleFetchStories();
	}, [handleFetchStories]);

	const handleSearchInput = (event) => {
		setSearchTerm(event.target.value);
	};
	const handleSearchSubmit = (event) => {
		setUrl(`${API_ENDPOINT}${searchTerm}`);
		event.preventDefault();
	};

	// If the App component re-renders, it always creates a new version of this callback handler as a new function.
	// Earlier, we used Reactâ€™s useCallback Hook to prevent this behavior, by creating a function only on a re-render
	// (if one of its dependencies has changed):
	const handleRemoveStory = useCallback((item) => {
		dispatchStories({
			type: "REMOVE_STORY",
			payload: item,
		});
	}, []);

	console.log("B:App");
	return (
		<StyledContainer>
			<StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary>
			<SearchForm
				searchTerm={searchTerm}
				onSearchInput={handleSearchInput}
				onSearchSubmit={handleSearchSubmit}
			/>

			{stories.isError && <p>Something went wrong :/</p>}
			{stories.isLoading ? (
				<p>Loading ...</p>
			) : (
				<List list={stories.data} onRemoveItem={handleRemoveStory} />
			)}
		</StyledContainer>
	);
};

export default App;
