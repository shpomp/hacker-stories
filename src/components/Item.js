import { ReactComponent as Check } from "../check.svg";

const Item = ({ item, onRemoveItem }) => (
	<li style={{ display: "flex" }}>
		<span style={{ width: "40%" }}>
			<a href={item.url}>{item.title}</a>
		</span>
		<span style={{ width: "30%" }}>{item.author}</span>
		<span style={{ width: "15%" }}>{item.num_comments}</span>
		<span style={{ width: "10%" }}>{item.points}</span>
		<span style={{ width: "5%" }}>
			<button
				className="button buttonSmall"
				type="button"
				onClick={() => onRemoveItem(item)}
			>
				<Check height="18px" width="18px" />
			</button>
		</span>
	</li>
);

export default Item;
