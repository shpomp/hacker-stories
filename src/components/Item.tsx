import { ReactComponent as Check } from "../check.svg";
import type { Story } from "../App";

type ItemProps = {
	item: Story;
	onRemoveItem: (item: Story) => void;
};

const Item = ({ item, onRemoveItem }: ItemProps) => (
	<li className="item">
		<span>
			<a href={item.url}>{item.title}</a>
		</span>
		<span>{item.author}</span>
		<span>{item.num_comments}</span>
		<span>{item.points}</span>
		<span>
			<button
				className="button buttonSmall"
				type="button"
				onClick={() => onRemoveItem(item)}
			>
				<Check height="18px" width="18px" />
			</button>
		</span>
	</li>
);

export default Item;
