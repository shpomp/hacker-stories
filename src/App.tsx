import axios from "axios";
import {
	useState,
	useEffect,
	useReducer,
	useCallback,
	useRef,
	useMemo,
} from "react";
import List from "./components/List";
import SearchForm from "./components/SearchForm";
import "./App.css";

export type Story = {
	objectID: string;
	url: string;
	title: string;
	author: string;
	num_comments: number;
	points: number;
};
const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

const useSemiPersistentState = (
	key: string,
	initialState: string
): [string, (newValue: string) => void] => {
	const isMounted = useRef(false);
	const [value, setValue] = useState(localStorage.getItem(key) || initialState);
	useEffect(() => {
		if (!isMounted.current) {
			isMounted.current = true;
		} else {
			localStorage.setItem(key, value);
		}
	}, [value, key]);
	return [value, setValue];
};

const ACTIONS = {
	SET_STORIES: "SET_STORIES",
	REMOVE_STORY: "REMOVE_STORY",
	STORIES_FETCH_INIT: "STORIES_FETCH_INIT",
	STORIES_FETCH_SUCCESS: "STORIES_FETCH_SUCCESS",
	STORIES_FETCH_FAILURE: "STORIES_FETCH_FAILURE",
};

const storiesReducer = (state, action) => {
	switch (action.type) {
		case ACTIONS.STORIES_FETCH_INIT:
			return {
				...state,
				isLoading: true,
				isError: false,
			};
		case ACTIONS.STORIES_FETCH_SUCCESS:
			return {
				...state,
				isLoading: false,
				isError: false,
				data: action.payload,
			};
		case ACTIONS.STORIES_FETCH_FAILURE:
			return {
				...state,
				isLoading: false,
				isError: true,
			};
		case ACTIONS.REMOVE_STORY:
			return {
				...state,
				data: state.data.filter(
					(story) => action.payload.objectID !== story.objectID
				),
			};
		default:
			throw new Error();
	}
};

const getSumComments = (stories) => {
	return stories.data.reduce((result, value) => result + value.num_comments, 0);
};

const App = () => {
	const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
	const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);

	const [stories, dispatchStories] = useReducer(storiesReducer, {
		data: [],
		isLoading: false,
		isError: false,
	});

	const handleFetchStories = useCallback(async () => {
		dispatchStories({ type: ACTIONS.STORIES_FETCH_INIT });

		try {
			const result = await axios.get(url);

			dispatchStories({
				type: ACTIONS.STORIES_FETCH_SUCCESS,
				payload: result.data.hits,
			});
		} catch {
			dispatchStories({ type: ACTIONS.STORIES_FETCH_FAILURE });
		}
	}, [url]);

	useEffect(() => {
		handleFetchStories();
	}, [handleFetchStories]);

	const handleSearchInput = (event) => {
		setSearchTerm(event.target.value);
	};

	const handleSearchSubmit = (event) => {
		setUrl(`${API_ENDPOINT}${searchTerm}`);

		event.preventDefault();
	};

	// If the App component re-renders, it always creates a new version of this callback handler as a new function.
	// Earlier, we used React’s useCallback Hook to prevent this behavior, by creating a function only on a re-render
	// (if one of its dependencies has changed):
	const handleRemoveStory = (item: Story) => {
		dispatchStories({
			type: "REMOVE_STORY",
			payload: item,
		});
	};

	//We can tell React to only run a function if one of its dependencies has changed.
	//If no dependency changed, the result of the function stays the same. React’s useMemo Hook helps us here:
	const sumComments = useMemo(() => getSumComments(stories), [stories]);

	return (
		<div className="container">
			<h1 className="headlinePrimary">
				My Hacker Stories with {sumComments} comments.
			</h1>
			<SearchForm
				searchTerm={searchTerm}
				onSearchInput={handleSearchInput}
				onSearchSubmit={handleSearchSubmit}
			/>

			{stories.isError && <p>Something went wrong :/</p>}
			{stories.isLoading ? (
				<p>Loading ...</p>
			) : (
				<List list={stories.data} onRemoveItem={handleRemoveStory} />
			)}
		</div>
	);
};

export default App;
