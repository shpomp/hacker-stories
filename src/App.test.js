import App, { storiesReducer } from "./App";
import { render, screen, fireEvent, act } from "@testing-library/react";
import Item from "./components/Item";

describe("something truthy and falsy", () => {
	test("true to be true", () => {
		expect(true).toBeTruthy();
	});

	test("false to be false", () => {
		expect(false).toBeFalsy();
	});
});

const storyOne = {
	title: "React",
	url: "https://reactjs.org/",
	author: "Jordan Walke",
	num_comments: 3,
	points: 4,
	objectID: 0,
};

const storyTwo = {
	title: "Redux",
	url: "https://redux.js.org/",
	author: "Dan Abramov, Andrew Clark",
	num_comments: 2,
	points: 5,
	objectID: 1,
};

const stories = [storyOne, storyTwo];

describe("storiesReducer", () => {
	test("removes a story from all stories", () => {
		const action = { type: "REMOVE_STORY", payload: storyOne };
		const state = { data: stories, isLoading: false, isError: false };
		const newState = storiesReducer(state, action);
		const expectedState = {
			data: [storyTwo],
			isLoading: false,
			isError: false,
		};
		expect(newState).toStrictEqual(expectedState);
	});
});

// UNIT

describe("Item", () => {
	test("renders all properties", () => {
		render(<Item item={storyOne} />);
		// screen.debug(); // The function gives a useful overview of what is rendered and informs the best way to proceed with testing.
		expect(screen.getByText("Jordan Walke")).toBeInTheDocument();
		expect(screen.getByText("React")).toHaveAttribute(
			"href",
			"https://reactjs.org/"
		);
	});

	test("renders a clickable dismiss button", () => {
		render(<Item item={storyOne} />);
		expect(screen.getByRole("button")).toBeInTheDocument();
	});

	test("clicking the dismiss button calls the callback handler", () => {
		const handleRemoveItem = jest.fn();
		render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);
		fireEvent.click(screen.getByRole("button"));
		expect(handleRemoveItem).toHaveBeenCalledTimes(1);
	});
});
