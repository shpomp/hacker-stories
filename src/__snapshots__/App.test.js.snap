// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<li
  class="item"
>
  <span>
    <a />
  </span>
  <span />
  <span />
  <span />
  <span>
    <button
      class="button buttonSmall"
      type="button"
    >
      dismiss
      <svg
        height="18px"
        width="18px"
      >
        check.svg
      </svg>
    </button>
  </span>
</li>
`;

exports[`SearchForm renders snapshot 1`] = `
<form
  class="searchForm"
>
  <label
    class="label"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button buttonLarge"
    type="submit"
  >
    Submit
  </button>
</form>
`;
